Password required dialog (web&native) and native room lock using basic react dialogs.

damencho opened the pull request:
Password required dialog and room lock feature uses the base dialogs.

damencho commented:
Jenkins test this please.

damencho commented:
Jenkins, test this please!

damencho commented:
@saghul and @hristoterezov Could you please take a look only at the first commit. For the others we haven't decided yet, cause there will be different styling for dialogs and we will merge them or work on them later.

hristoterezov commented on react/features/base/dialog/actionTypes.js:
```diff
@@ -0,0 +1,22 @@
+ * }
+ *
+ */
+export const BEGIN_DIALOG_REQUEST = Symbol('BEGIN_DIALOG_REQUEST');
```
Isn't it better if the actionType names are similar or the same as the actions?

damencho replied:
@saghul I was using the names that were used in native, what do you think about it? What about renaming the action types to SHOW_DIALOG and HIDE_DIALOG, wdyt?

saghul replied:
I think in this case we do want a begin...end sequence. Imagine a dialog where user input is requested, or where multiple buttons are presented, the ended action could contain data about that result.

hristoterezov replied:
@saghul I don't think I understand your argument. We are talking about renaming only? For me it make sense to have default("abstract") actions for dialog open and close because there isn't anything else you can do with the dialog on this generic level.

hristoterezov commented on react/features/base/dialog/components/Dialog.web.js:
```diff
@@ -0,0 +1,212 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import { translate } from '../../i18n';
```
I think we are leaving empty row between the 3rd party imports and our modules.

hristoterezov commented on react/features/base/dialog/components/Dialog.web.js:
```diff
@@ -0,0 +1,212 @@
+import ModalDialog from '@atlaskit/modal-dialog';
+import AKButtonGroup from '@atlaskit/button-group';
+import AKButton from '@atlaskit/button';
+
```
This empty line should be removed.

hristoterezov commented on react/features/base/dialog/components/Dialog.web.js:
```diff
@@ -0,0 +1,212 @@
+
+import ModalDialog from '@atlaskit/modal-dialog';
+import AKButtonGroup from '@atlaskit/button-group';
+import AKButton from '@atlaskit/button';
```
Since those are 3rd party imports they should go at the beginning of the file.

hristoterezov commented on react/features/base/dialog/components/Dialog.web.js:
```diff
@@ -0,0 +1,212 @@
+import React, { Component } from 'react';
```
Every group of imports should be alphabetically ordered.

hristoterezov commented on react/features/base/dialog/components/Dialog.web.js:
```diff
@@ -0,0 +1,212 @@
+     * @static
+     */
+    static propTypes = {
+        cancelDisabled: React.PropTypes.bool,
```
Missing JSDoc for the properties.

hristoterezov commented on react/features/base/dialog/components/DialogContainer.js:
```diff
@@ -0,0 +1,54 @@
+     * @static
+     */
+    static propTypes = {
+        _component: React.PropTypes.func,
```
Missing JSDoc

hristoterezov commented on react/features/base/dialog/reducer.js:
```diff
@@ -0,0 +1,33 @@
+    END_DIALOG_REQUEST
+} from './actionTypes';
+
+import { ReducerRegistry, setStateProperties } from '../redux';
```
The imports from the other modules should be ordered before the imports from the same module.

damencho commented:
@hristoterezov, I think I've addressed all your comments except the one about actionType names, the one I asked @saghul about.

saghul commented on react/features/base/dialog/components/Dialog.web.js:
```diff
@@ -0,0 +1,267 @@
+/**
+ * Web dialog that uses atlaskit modal-dialog to display dialogs.
+ */
+class Dialog extends Component {
```
There seems to be some duplication here, how about creating an AbstractDialog and putting the common bits there?

saghul commented on react/features/base/dialog/components/Dialog.native.js:
```diff
@@ -0,0 +1,127 @@
+        cancelTitleKey: React.PropTypes.string,
+        dispatch: React.PropTypes.func,
+        onCancel: React.PropTypes.func,
+        onSubmit: React.PropTypes.func,
```
This suggests dialogs will always have some submit and cancel buttons, what if we need more? I'd like to get an array ob objects. Each object represents a button, with its text + the associated handler function. WDYT?

damencho replied:
This is something @yanas was insisting on: to have a common dialog style with one, two or no buttons.

saghul commented on react/features/base/dialog/components/Dialog.native.js:
```diff
@@ -0,0 +1,127 @@
+     * @returns {ReactElement}
+     */
+    render() {
+        const { t } = this.props;
```
this could be added to de destructure below

saghul commented on react/features/base/dialog/components/Dialog.web.js:
```diff
@@ -0,0 +1,267 @@
+         * Whether the dialog is modal. This means clicking on the blanket will
+         * leave the dialog open. No cancel button.
+         */
+        isModal: React.PropTypes.bool,
```
On mobile all dialogs are modal. Do we have an instance of non modal dialogs on the web?

damencho replied:
Yep like sharing youtube video is not modal, you can click outside the dialog and this will dismiss it and you can continue your work, the same goes for recording maybe and others.

saghul commented on react/features/base/dialog/reducer.js:
```diff
@@ -0,0 +1,33 @@
+import { ReducerRegistry, setStateProperties } from '../redux';
```
import order

damencho replied:
Seems fine to me. 2 imports, one is ours '../redux', the other one is separate by newline as it is from the same feature './actionTypes'.

saghul replied:
I think the first one should be the actionTypes

saghul commented on react/features/base/dialog/reducer.js:
```diff
@@ -0,0 +1,33 @@
+
+    switch (action.type) {
+    case BEGIN_DIALOG_REQUEST:
+
```
why the newline?

saghul commented on react/features/base/dialog/reducer.js:
```diff
@@ -0,0 +1,33 @@
+        });
+    case END_DIALOG_REQUEST:
+        return setStateProperties(state, {
+            show: false
```
we should cleanup the component and props

saghul commented on react/features/room-lock/actions.js:
```diff
@@ -19,10 +19,7 @@ export function beginRoomLockRequest(conference) {
-                type: BEGIN_ROOM_LOCK_REQUEST,
-                conference
-            });
+            dispatch(openDialog(RoomLockPrompt, { conference }));
```
❤️

saghul commented on react/features/room-lock/actions.js:
```diff
@@ -43,13 +40,23 @@ export function endRoomLockRequest(conference, password) {
+ * @protected
+ * @returns {{
+ *     type: BEGIN_DIALOG_REQUEST
+ * }}
```
complete the docs

saghul commented on react/features/room-lock/components/PasswordRequiredPrompt.web.js:
```diff
@@ -0,0 +1,127 @@
+import { translate } from '../../base/i18n';
+import { Dialog } from '../../base/dialog';
+
+import AKFieldText from '@atlaskit/field-text';
```
order

damencho commented:
Hey @saghul and @hristoterezov, I just force pushed updating with the comments I had from you both.
So I moved Dialog.native from third commit to the first one and introduced AbstractDialog sharing the common logic between web and native. For now, the goal is to merge only the first commit, so we do not block work that is dialog related.

hristoterezov commented on react/features/base/dialog/components/AbstractDialog.js:
```diff
@@ -0,0 +1,133 @@
+    const dialog = state['features/base/dialog'];
+
+    return {
+        show: dialog.show
```
I think this property should be named with "_" prefix since it's coming from mapStateToProps.

damencho replied:
But this show is used in Dialog.web and Dialog.native

saghul commented on react/features/base/dialog/components/Dialog.web.js:
```diff
@@ -0,0 +1,161 @@
+                header = { this._renderHeader() }
+                isOpen = { this.props.show }
+                onDialogDismissed = { this._onCancel }
+                width = { this.props.width
```
this.props.width || 'medium'

saghul commented on react/features/base/dialog/components/Dialog.web.js:
```diff
@@ -0,0 +1,161 @@
+     * @private
+     */
+    _renderCancelButton() {
+        if (this.props.cancelDisabled || this.props.isModal) {
```
so modal dialogs don't have a cancel button? ever?

damencho replied:
Yes. It doesn't work if we have. The thing is that clicking outside dialog is the same as clicking cancel, so modality will not work if having a cancel. This is limitation of atlaskit modal dialog. I discussed this with @yanas and she agreed with this approach.

saghul commented on react/features/base/dialog/components/Dialog.web.js:
```diff
@@ -0,0 +1,161 @@
+                id = 'modal-dialog-cancel-button'
+                onClick = { this._onCancel }>
+                { this.props.t(this.props.cancelTitleKey
+                    ? this.props.cancelTitleKey : 'dialog.Cancel') }
```
same as before, use ||

saghul commented on react/features/base/dialog/components/Dialog.web.js:
```diff
@@ -0,0 +1,161 @@
+                id = 'modal-dialog-submit-button'
+                onClick = { this._onSubmit }>
+                { this.props.t(this.props.submitTitleKey
+                    ? this.props.submitTitleKey : 'dialog.Ok') }
```
ditto

saghul commented on react/features/base/dialog/components/DialogContainer.js:
```diff
@@ -0,0 +1,62 @@
+        /**
+         * The component to render.
+         */
+        _component: React.PropTypes.func,
```
why are these private props?

damencho replied:
Is there some rule about making props private? I see you and @hristoterezov are on a different opinion for show and _show.

damencho replied:
Internally used only by the container ...

saghul commented on react/features/base/dialog/components/DialogContainer.js:
```diff
@@ -0,0 +1,62 @@
+     * @returns {ReactElement}
+     */
+    render() {
+
```
extra newline

saghul commented on react/features/base/dialog/components/DialogContainer.js:
```diff
@@ -0,0 +1,62 @@
+        /**
+         * The props to pass to the component that will be rendered.
+         */
+        _props: React.PropTypes.object
```
this reads a bit weird. maybe we can rename this to componentProps ?

saghul commented on react/features/base/dialog/components/DialogContainer.js:
```diff
@@ -0,0 +1,62 @@
+ * }}
+ */
+function _mapStateToProps(state) {
+
```
extra newline

saghul commented on react/features/base/dialog/reducer.js:
```diff
@@ -0,0 +1,34 @@
+ * @returns {{}}
+ */
+ReducerRegistry.register('features/base/dialog', (state = {}, action) => {
+
```
extra newline

damencho commented on react/features/base/dialog/components/AbstractDialog.js:
```diff
@@ -0,0 +1,133 @@
+        /**
+         * Is submit button enabled/disabled. Enabled by default.
+         */
+        submitDisabled: React.PropTypes.bool,
```
I was adding it as we had a disable for the other one ... Maybe having only cancel button, this were my thoughts.

damencho commented on react/features/base/dialog/actions.js:
```diff
@@ -0,0 +1,32 @@
+ */
+export function hideDialog() {
+    return {
+        type: END_DIALOG_REQUEST
```
Yep, that's the reason to have the container so we can play with multliple dialogs if need to. I think I have seen this in current meet.

hristoterezov commented on react/features/base/dialog/components/Dialog.native.js:
```diff
@@ -0,0 +1,58 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import { translate } from '../../i18n';
```
Missing new line before this import.

hristoterezov commented on react/features/base/dialog/components/Dialog.native.js:
```diff
@@ -0,0 +1,58 @@
+import { connect } from 'react-redux';
+import { translate } from '../../i18n';
+
+import Prompt from 'react-native-prompt';
```
This import should be in the first group of imports for the 3rd party modules, right?

saghul commented on react/features/room-lock/middleware.js:
```diff
@@ -0,0 +1,36 @@
+MiddlewareRegistry.register(store => next => action => {
+
+    switch (action.type) {
+    case CONFERENCE_FAILED: {
```
what happens if the conference fails while the password prompt is visible? is this a real possibility?

damencho replied:
I don't think this is possible, seeing password dialog means you are not in a muc, so technically nothing is going on.

This PR, at least the web part, will not be merged as it will lead to inconsistency in UI.

saghul commented:
> This PR, at least the web part, will not be merged as it will lead to inconsistency in UI.

I'm not sure what that means. Can you clarify @damencho ?

damencho commented:
@saghul , we do not want to merge the web part as there will be dialogs with different styles. We haven't discuss it, how to proceed to avoid this, other than having everything rectified.

saghul commented:
@damencho 👍
